### PR TITLE
Reorder beacon advertising procedure and better error reporting.

### PR DESCRIPTION
**Pull request by tomasbrod**
BOINCer joined #troubleshooting by invitation from .
BOINCer 10:53 PM
cpid is af73cd19f79a0ade8e6ef695faa2c776
vega40k 10:54 PM
joined #troubleshooting.
brod 10:58 PM

BEACON add af73cd19f79a0ade8e6ef695faa2c776 af73cd19f79a0ade8e6ef695faa2c77666663bc83d349941c694369636c7646cc543966a3a973ec49e649e3b696dca966f70716274626f6265624171737075706f6e626a6d2f64706e;5210f125612c2237ae69b20525c6dc289e00b6c4d84ac643c3079a151fc8bfa5;S2fbFMoGeBJFGsuZFakE2F8bcxF7yiUFkX;04769d0cdbadd8665b1b5f51aa9fdfc6fe8da223948bace367391549eb651bd50fef43e3b3df32432a4722100f17b1dbf61bdf35f3df60537a26778c124a60c66a 10-01-2017 03:43:15

BEACON DEL af73cd19f79a0ade8e6ef695faa2c776 - 10-06-2017 16:30:10

BEACON add af73cd19f79a0ade8e6ef695faa2c776 af73cd19f79a0ade8e6ef695faa2c776c46a97673a973a9e659632943fc96acb669a386d38943b653e633d38686f70386f70716274626f6265624171737075706f6e626a6d2f64706e;3b95f228a4d339af3d9f014684b142e35538e8fb66de183b0c95cf03fc0d34d7;S5rFS27syvbz9hv8qeDsaDSTZZobHkeavL;0477ab60661cf78e2b216e82977a987fe8062754dd7962341a100b32bdfeb9fe5e628f2fe91ced9bb411d555ab18e3659f8e8446bc054ed4255ba87ce6d8d3bcb6 10-06-2017 18:53:21

BEACON DEL af73cd19f79a0ade8e6ef695faa2c776 - 10-07-2017 00:48:51

BEACON add af73cd19f79a0ade8e6ef695faa2c776 af73cd19f79a0ade8e6ef695faa2c776c4c339c73c393b426938333c9ac8c96f6941956c94343ac63e673d9b69cb6c336f70716274626f6265624171737075706f6e626a6d2f64706e;a17190c2d31752a3ad42d842342d951170ab2eb0afe79d140f45cb0f9e31fa34;S5rFS27syvbz9hv8qeDsaDSTZZobHkeavL;04de362e82efb25b40683ec29cc06039021dddf7c14ffdd5197a71b3e2b30d5cbe62a4dd48b66b20296dec8c37a77e8abe7610bd20b6343cad21752c174c7b8d9a 10-07-2017 01:15:50

BEACON DEL af73cd19f79a0ade8e6ef695faa2c776 - 10-07-2017 18:32:49

BEACON add af73cd19f79a0ade8e6ef695faa2c776 af73cd19f79a0ade8e6ef695faa2c7766a663aca3d3c9c3dc695953d3fc66a676a41966f363b9a66436b3c98cc6cca956f70716274626f6265624171737075706f6e626a6d2f64706e;22adb862b20d1f448f2b28b41ab7276dd4b83962a6917ea47dca52ffa1f29863;S5rFS27syvbz9hv8qeDsaDSTZZobHkeavL;042df684ccb8b69c23e72064813cd8c533ffe4c7e942897639c8f36b9b32453b59e53178825efc1bf6affbbee7ab655f81ffde816edcc6922aa1dfd064a6bc0364 10-07-2017 19:35:46

BEACON add af73cd19f79a0ade8e6ef695faa2c776 af73cd19f79a0ade8e6ef695faa2c7766a663aca3d3c9c3dc695953d3fc66a676a41966f363b9a66436b3c98cc6cca956f70716274626f6265624171737075706f6e626a6d2f64706e;22adb862b20d1f448f2b28b41ab7276dd4b83962a6917ea47dca52ffa1f29863;S5rFS27syvbz9hv8qeDsaDSTZZobHkeavL;042df684ccb8b69c23e72064813cd8c533ffe4c7e942897639c8f36b9b32453b59e53178825efc1bf6affbbee7ab655f81ffde816edcc6922aa1dfd064a6bc0364 10-07-2017 19:35:46

brod 10:59 PM
OK, these are your beacon transactions. I see you deleted them a lot. @BOINCer
10:59
the last appears to be from today
11:00
do not worry about the Address in beacon, it is unused
BOINCer 11:01 PM
yeah, i tried today to delete deacons from all my known grc addresses, transfer all coins to new address, then advertise.  but i get wallet unlock errors, even though the wallet is unlocked and beacon transaction goes out. cpid in beaconreport is listing grc address with no coins in it, not the grc address that sent the beacon trasaction
brod 11:02 PM
Your public key is 042df684ccb8b69c23e72064813cd8c533ffe4c7e942897639c8f36b9b32453b59e53178825efc1bf6affbbee7ab655f81ffde816edcc6922aa1dfd064a6bc0364, check if you have that in config. Do NOT modify anything yet
BOINCer 11:02 PM
here's my execute beaconstatus:
11:02
14:02:14
￼
execute beaconstatus

14:02:14
￼
[
{
"Command" : "beaconstatus"
},
{
"CPID" : "af73cd19f79a0ade8e6ef695faa2c776",
"Beacon Exists" : "Yes",
"Beacon Timestamp" : "10-07-2017 19:35:46",
"Public Key" : "042df684ccb8b69c23e72064813cd8c533ffe4c7e942897639c8f36b9b32453b59e53178825efc1bf6affbbee7ab655f81ffde816edcc6922aa1dfd064a6bc0364",
"Private Key" : "",
"Local Configuration Public Key" : "",
"Magnitude (As of last superblock)" : 0.00000000,
"Warning" : "Your magnitude is 0 as of the last superblock: this may keep you from staking POR blocks.",
"Block Signing Test Results" : false,
"Errors" : "Private Key Missing. Local configuration file Public Key missing. Local configuration public key does not match beacon public key. This can happen if you copied the wrong public key into your configuration file. Please request that your beacon is deleted, or look into walletbackups for the correct keypair. Failed to sign POR block. This can happen if your keypair is invalid. Check walletbackups for the correct keypair, or request that your beacon is deleted. ",
"Help" : "Note: If your beacon is missing its public key, or is not in the chain, you may try: execute advertisebeacon.",
"Configuration Status" : "FAIL"
}
]
11:02
public key is not in .conf file
brod 11:02 PM
the line should read `publickeyaf73cd19f79a0ade8e6ef695faa2c776=042df684ccb8b69c23e72064813cd8c533ffe4c7e942897639c8f36b9b32453b59e53178825efc1bf6affbbee7ab655f81ffde816edcc6922aa1dfd064a6bc0364`
11:03
do NOT change it
11:03
@jringo listen to this
BOINCer 11:03 PM
thats what i thought, but this is not in .conf file, just email, cpid, and nodes
brod 11:04 PM
@ifoggz Your config editing code does not work again, argh!
11:04
Now go to the walletbackups directory
BOINCer 11:05 PM
ok, i can get keys from there but they have my old cpid prefix and public key does not match this one
brod 11:05 PM
look in all the conf backups and try to see if this public key is in any of them
BOINCer 11:07 PM
no dice.  all are using a different key from my old cpid.
brod 11:09 PM
It seems the backup procedure is broken. Even my wallet auto-advertised today, and created NO backups. @ifoggz
BOINCer 11:10 PM
yeah if i try to backup from the gui it crashes on windows 10
brod 11:12 PM
@BOINCer Ok. Then delete the beacon again. Clean your config of all keys and cpids. Start up wallet again, check beaconstatus to see if it detected your cpid right, only then unlock your wallet fully and execute advertisebeacon. Just in case copy/save the output. And restart wallet.
11:12
@quezacoatl can you delete this poor guy's beacon again? CPID=af73cd19f79a0ade8e6ef695faa2c776 (edited)
BOINCer 11:12 PM
NOOO not my bacon!
11:13
I love bacon
brod 11:13 PM
Also @BOINCer There is no need to request beacon delete if you CHANGE CPID.
BOINCer 11:13 PM
I think Quez just went to sleep, i've been emailing him
brod 11:13 PM
I can do it. wait
BOINCer 11:15 PM
waiting on transaction confirmation.  one weird thing i noticed before is that the beaconreport showed my cpid with a blank grc address, even 16 hours after deleting the beacon a few days ago
brod 11:15 PM
that grc address has no role in beacon. it is there from historic reasons and is not used for anything
BOINCer 11:16 PM
hmm, beacon deletion is rejected (0.00011 back to me)
brod 11:16 PM
stop doing anything
11:16
i am working on it
BOINCer 11:17 PM
:slightly_smiling_face: thanks
brod 11:18 PM
<MT>beacon</MT><MK>af73cd19f79a0ade8e6ef695faa2c776</MK><MV>1</MV><MA>D</MA><MPK></MPK><MS>MEQCIBuCO+dm9hCqYyUhXmu4AZEIZs299HooIzT1nz1ps3u9AiBr8f86gbbiHhiOc7S4Mu43Js7rHmvT121fxlHW5pqJCg==</MS>
BOINCer 11:18 PM
totally (?) :slightly_smiling_face:
brod 11:20 PM
Sent: a4d6c81880c190a6d955b0bdd642e5bd5b10a2adf5cb5e094885d06ca2da1e2d
11:21
Sorry for the gibberish. I used this as clipboard. Your beacon should be deleted in 10 blocks.
11:21
Please lock your wallet so it does not try to make a new one too soon.
BOINCer 11:21 PM
wallet is locked!
brod 11:21 PM
Clean your config of all keys and cpids. Better wait about 20 minutes. Start up wallet again, check beaconstatus to see if it detected your cpid right, only then unlock your wallet fully and execute advertisebeacon. Just in case copy/save the output. And restart wallet. (edited)
BOINCer 11:24 PM
remove nodes except for node.gridcoin.us?  i have a bunch in there
11:25
OK thanks again for your help.  I'm gonna go get drunk at a bbq so that should be enough time to pass 10 blocks :stuck_out_tongue:
brod 11:25 PM
It is deleted now.
BOINCer 11:26 PM
so... think i should try now?
brod 11:26 PM
no, keep the nodes in there, as well as any paths and option you set
11:26
yes, You can try now
BOINCer 11:27 PM
no paths set, default boinc install. there was a neural network entry in there but i deleted it.
11:27
cool here we go!
11:28
run as administrator, check
brod 11:29 PM
I will watch the network logs. @brod me when you sent
11:31
Your logs could be useful in this case, but the wallet overwrites them on startup :disappointed:
BOINCer 11:31 PM
d'oh
11:32
sheet
theearl 11:32 PM
hacker
BOINCer 11:33 PM
so any advice for unlocking? I get errors in advertisebeacon even when unlocking from debug, and transaction goes through...
11:33
so that's weird
11:34
so unlock from GUI or debug is better?
brod 11:34 PM
transaction was sent, don't mess with it (edited)
BOINCer 11:35 PM
cpid good in beaconstatus
11:35
ok, i did not unlock wallet or advertise...?
11:36
no transaction showing in my gui log
11:37
beaconreport shows my ssid with blank grc field
brod 11:37 PM
Oh, I thought you sent already. The go for it, unlock and send
BOINCer 11:38 PM
is gui or debug lock any different?
brod 11:38 PM
no, they are same
BOINCer 11:39 PM
OK here we go!
11:39
￼
[
{
"Command" : "advertisebeacon"
},
{
"Result" : "FAIL",
"CPID" : "af73cd19f79a0ade8e6ef695faa2c776",
"Message" : "6984f75f192b27d9e376eeb0e3802017ca01d71378ba9fecaec0f90146f5ca1b",
"Errors" : "Error: Unable to send beacon::Wallet Locked::Please enter the wallet passphrase with walletpassphrase first.",
"FAILURE" : "Note: if your wallet is locked this command will fail; to solve that unlock the wallet: 'walletpassphrase <yourpassword> <240>'."
}
]
11:40
but wallet is unlocked and transaction just went out
brod 11:40 PM
WTF? I just saw and af73cd19f79a0ade8e6ef695faa2c776 beacon transaction
11:40
copy your debug.log to safe place
BOINCer 11:41 PM
yeah debug says the wallet is locked, but transaction goes out...?
brod 11:41 PM
You can upload it here
11:42
what version are you running?
BOINCer 11:42 PM
3.6.2.0-unk
brod 11:42 PM
latest, hmmm. Upload the log here
BOINCer 11:44 PM
ok how to upload, and which is log file? deebug? debug2?
11:44
still at 2/10 confirmations...
11:45
ok to lock wallet?
11:46
14:45:59
￼
execute beaconstatus

14:45:59
￼
[
{
"Command" : "beaconstatus"
},
{
"CPID" : "af73cd19f79a0ade8e6ef695faa2c776",
"Beacon Exists" : "Yes",
"Beacon Timestamp" : "10-07-2017 21:35:04",
"Public Key" : "04da2e8fc745583a29f6f137c97f880dc2b42948ab52ecd89ca50b3029cc2bf310617aa97d5f5d203ae1f7e7a9bf2ef76a844d73a018d2277f73323025ed766132",
"Private Key" : "",
"Local Configuration Public Key" : "",
"Magnitude (As of last superblock)" : 0.00000000,
"Warning" : "Your magnitude is 0 as of the last superblock: this may keep you from staking POR blocks.",
"Block Signing Test Results" : false,
"Errors" : "Private Key Missing. Local configuration file Public Key missing. Local configuration public key does not match beacon public key. This can happen if you copied the wrong public key into your configuration file. Please request that your beacon is deleted, or look into walletbackups for the correct keypair. Failed to sign POR block. This can happen if your keypair is invalid. Check walletbackups for the correct keypair, or request that your beacon is deleted. ",
"Help" : "Note: If your beacon is missing its public key, or is not in the chain, you may try: execute advertisebeacon.",
"Configuration Status" : "FAIL"
}
]
11:46
same old story
brod 11:47 PM
`debug.log`
11:47
Your new public key is 04da2e8fc745583a29f6f137c97f880dc2b42948ab52ecd89ca50b3029cc2bf
310617aa97d5f5d203ae1f7e7a9bf2ef76a844d73a018d2277f73323025ed766132
11:47
upload the file with the plus on the left
BOINCer 11:49 PM
debug.log

2MB Plain Text Click to download
BOINCer 11:49 PM
voila
11:53
ok i will be back in a few hours. thanks again for your help, looking forward to getting started!!
brod 11:54 PM
No error in the logs.
BOINCer 11:54 PM
kewl
11:55
so it's a mystery why the keys aren't getting put in the .conf file
11:55
to me anyway :slightly_smiling_face:
brod 11:56 PM
It is not good, if there was error we would know what went wrong. I will delete your beacon again, obviously your next try is going same way. This is definitely a bug. We will have a look. Thank you cooperation.
11:56
sent 26f19c1bd019dced10d20a7bc882e6bdae6949ed7afb6874e31ea310f3f82db9
BOINCer 11:57 PM
*sigh* well thanks, i thought i was doing something wrong.  thanks for your help.
11:57
figured if i didn't stake a block in 5 days with 900000 coin weight something must be wrong.
brod 11:58 PM
If you want interest, you can go into INVESTOR mode (config `boincdatadir=/tmp/non_existent_directory`)
BOINCer 11:59 PM
nah, just want to get some por credit for my crunching farm for now. but thanks.
11:59
wallet is locked until i hear back from you.
Today
brod 12:00 AM
check out `getmininginfo`, shows what you are oved
jringo 12:00 AM
listened and learned
12:00
ty
BOINCer 12:01 AM
if i were to change boinc email/cpid and try again, do you think i'll have the same problem?
12:03
0 owed, as i can't advertise beacon correctly (i think?)  should be about 100 grc by now, but that's like $3.  much less than the 15-20 hours of troubleshooting so far
brod 12:03 AM
I wis it would wotk
BOINCer 12:03 AM
:slightly_smiling_face: i know, thanks again.  you know what does work?  :beer:
12:04
just learned how to post emojis :dromedary_camel:
12:04
:poop:
brod 12:04 AM
i think i found the issue
12:05
cancer

### CODE REVIEW

**Comment by iFoggz** [comment]
crap the backups didnt work because we change the way the directory was done in Backup. GetBackupFileName appends backuppath itself now. my bad. and i like the extra error messages but most of the AddContract code was previous code that I put the backup related stuff in there into. so i broke the backups on this one. I thought I did go over this but now I realize i did before i changed everything

**Comment by tomasbrod** [comment]
I changed the exception printing so it compiles.
If you feel it is too fragile, I can remove it.

File: src/rpcblockchain.cpp [hunk 1]
```diff
@@ -1007,36 +1007,44 @@ bool AdvertiseBeacon(std::string &sOutPrivKey, std::string &sOutPubKey, std::str
...
+                }
+
                 // Backup config with new keys with beacon suffix
-                std::string sBeaconBackupNewConfigFilename = GetBackupFilename("gridcoinresearch.conf", "beacon");
-                boost::filesystem::path sBeaconBackupNewConfigTarget = GetDataDir() / "walletbackups" / sBeaconBackupNewConfigFilename;
-                BackupConfigFile(sBeaconBackupNewConfigTarget.string().c_str());
+                StoreBeaconKeys(GlobalCPUMiningCPID.cpid, sOutPubKey, sOutPrivKey);
+                if(!BackupConfigFile(GetBackupFilename("gridcoinresearch.conf", "beacon")))
+                {
+                    sError = "Failed to backup new configuration file. Beacon not sent. And your config file is probably corrupted now.";
```
Comment (denravonska) [inline src/rpcblockchain.cpp]: Change this message to something more user friendly. 

> Failed to back up configuration file. Beacon not sent, please manually roll back to previous configuration.

File: src/rpcblockchain.cpp [hunk 2]
```diff
@@ -1007,36 +1007,44 @@ bool AdvertiseBeacon(std::string &sOutPrivKey, std::string &sOutPubKey, std::str
...
+                sError = "Error: Unable to send beacon::"+json_spirit::write_string(json_spirit::Value(objError),true);
                 return false;
             }
             catch (std::exception &e) 
             {
-                sError = "Error: Unable to send beacon::Wallet Locked::Please enter the wallet passphrase with walletpassphrase first.";
+                sError = "Error: Unable to send beacon;:"+std::string(e.what());
                 return false;
             }
             catch(...)
```
Comment (denravonska) [inline src/rpcblockchain.cpp]: Honestly, I think we can drop this `...`. We shouldn't be throwhing anything that isn't derived from std::exception.

File: src/rpcblockchain.cpp [hunk 3]
```diff
@@ -1007,36 +1007,44 @@ bool AdvertiseBeacon(std::string &sOutPrivKey, std::string &sOutPubKey, std::str
...
-                BackupConfigFile(sBeaconBackupNewConfigTarget.string().c_str());
+                StoreBeaconKeys(GlobalCPUMiningCPID.cpid, sOutPubKey, sOutPrivKey);
+                if(!BackupConfigFile(GetBackupFilename("gridcoinresearch.conf", "beacon")))
+                {
+                    sError = "Failed to backup new configuration file. Beacon not sent. And your config file is probably corrupted now.";
+                    return false;
+                }
+
+                // This prevents repeated beacons
+                nLastBeaconAdvertised = nBestHeight;
```
Comment (iFoggz) [inline src/rpcblockchain.cpp]: this should be below AddContract?. If AddContract fails then they will have to wait 5 blocks to do another attempt for instance if the wallet is locked.

File: src/kernel.cpp [hunk 4]
```diff
@@ -847,25 +855,12 @@ bool FindStakeModifierRev(uint64_t& nStakeModifier,CBlockIndex* pindexPrev)
...
     const CBlockIndex* pindex = pindexPrev;
-    const uint256 ModifGlitch_hash("12bcc37789ef00809d1287f4af3c46106181e7c60654c7d3f0677aefa8a78774");
-    const uint64_t ModifGlitch_correct=0xdf209a3032807577;
 
     while (1)
     {
         if(!pindex)
             return error("FindStakeModifierRev: no previous block from %d",pindexPrev->nHeight);
 
-        if (pindex->GetBlockHash()==ModifGlitch_hash)
```
Comment (denravonska) [inline src/kernel.cpp]: Was this supposed to be removed here?

**Pull request merged**